### PR TITLE
Introduce iterations & data feed (CSV/JSON) to address #5324, #5163

### LINKED ISSUE #5163: [feature]: Collection runner via GUI with ability to import files and choose which requests should be run
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Summary

Despite the fact that runner is already available via CLI and it allows users to provide input data it would be a huge improvement to be able to use it via GUI also like in Postman with same features. Another nice thing would be to choose which requests from collection should be run. 

### Why should this be worked on?

I believe it would improve overall experience and let users automate their tasks in more efficient ways

### LINKED ISSUE #5324: [feature]: Support setting number of iterations in Collection Runner UI
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Summary

Add a feature to the Hoppscotch Collection Runner UI to specify the number of iterations for a run, similar to the [existing functionality available in the CLI](https://docs.hoppscotch.io/documentation/clients/cli/overview#running-collections-present-on-the-api-client:~:text=no_of_iterations%3A%20Indicates%20the%20number%20of%20iterations%20to%20run%20the%20collection.%20Each%20iteration%20will%20run%20the%20entire%20collection%20once%2C%20replacing%20any%20iteration%2Dspecific%20data%20defined%20by%20the%20%2D%2Diteration%2Ddata%20flag%20(if%20provided).).

### Why should this be worked on?

The CLI already supports specifying the number of iterations when running a collection, which is great for automation and scripting. However, users who rely on the UI lack an intuitive way to repeat a collection run multiple times without manually re-triggering it.

Adding this to the UI would significantly improve usability and bring feature parity with the CLI. It would make the tool more friendly for less technical users and enable common workflows such as load testing, mass resource creation, and regression or flakiness testing through the UI.

### PR DESCRIPTION
**Pull request by Manohar-ford**
Screenshot 2025-11-04 102316
Screenshot 2025-11-04 102339
Screenshot 2025-11-04 102358

Closes #5324, Closes #5163

This Pull Request introduces significant enhancements to the Hoppscotch Collection Runner, enabling users to execute collections with simple iterations and drive request parameters using external data files. This allows for robust data-driven testing and automation of repetitive tasks, directly addressing user demands for more powerful collection execution capabilities.

### What's changed
This PR implements two major features within the Collection Runner: **Simple Iteration** and **Data Feed (CSV/JSON)**.

*   **Collection Runner - Simple Iteration:**
    *   A new input field has been added to the Collection Runner UI, allowing users to specify a fixed number of times a collection (or selected requests within it) should be executed.
    *   The runner will now loop through the entire collection `N` times, providing a straightforward way to repeat tests or actions.
    *   Each iteration will run with a fresh state, allowing for consistent testing across multiple runs.

*   **Collection Runner - Data Feed (CSV/JSON):**
    *   Users can now upload external data files (CSV or JSON format) to drive variables within their requests during collection runs.
    *   A new section in the Collection Runner UI allows for selecting and previewing the uploaded data file.
    *   For each row in the data file (or each object in a JSON array), a new iteration of the collection is executed.
    *   Data from the current row/object is automatically injected as environment variables for that specific iteration. For example, a CSV column `username` would be accessible as `{{username}}` in requests during that iteration.
    *   This enables dynamic population of request URLs, headers, body, and query parameters, facilitating data-driven testing scenarios.
    *   Robust parsing for both CSV and JSON formats is implemented, with clear feedback for malformed files.

*   **Integration and UI/UX:**
    *   The Collection Runner interface has been updated to include controls for both iteration count and data file upload.
    *   The execution logic of the Collection Runner has been refactored to manage the iteration loop and variable injection from data feeds.
    *   Improved feedback mechanisms during the run to indicate current iteration number and data context.

### Notes to reviewers
*   **Motivation:** This feature directly responds to the community's need for advanced testing capabilities, as highlighted in issues #5324 and #5163. It allows Hoppscotch to compete more effectively with other API clients in terms of automated testing and data-driven workflows.
*   **Testing Strategy:**
    *   **Manual Testing:** Extensive manual testing was performed to cover:
        *   Running collections with a specified number of iterations (e.g., 1, 5, 10).
        *   Uploading valid CSV and JSON files, ensuring correct variable injection into requests.
        *   Testing with various request types (GET, POST, PUT) and variable locations (URL, headers, body).
        *   Verifying error handling for malformed data files or missing variables.
        *   Confirming that environments are correctly handled per iteration.
    *   **Unit/Integration Tests:** Relevant unit and integration tests have been added/updated for the new iteration and data parsing logic.
*   **Edge Cases Considered:**
    *   Empty data files: Handled gracefully, resulting in zero iterations or an appropriate error message.
    *   Missing variables in data: Variables not found in the data file will fall back to existing environment/global variables or remain unresolved, consistent with current Hoppscotch behavior.
    *   Large data files: Performance considerations were made for parsing, though extremely large files might still impact initial load time.
    *   Data type handling: All data from CSV/JSON is treated as strings for variable substitution, allowing users to perform type conversions within their request scripts if needed.
*   **Future Enhancements (Out of Scope for this PR):**
    *   Advanced reporting for data-driven runs (e.g., pass/fail statistics per iteration).
    *   Support for additional data formats (e.g., Excel).
    *   More sophisticated iteration control (e.g., running specific rows, conditional iterations).
    *   Visual mapping of data file columns to request variables.
*   **Backward Compatibility:** This feature is fully backward compatible. Existing collection runs without iteration counts or data files will function as before.
*   **Security Considerations:** Data files are processed client-side. Users are responsible for the content of their uploaded files. No new external network requests are made as part of the data parsing.

### Checklist
*   [x] Linked issues (#5324, #5163)
*   [x] Backward compatible
*   [x] New UI elements for iteration count and data file upload
*   [x] Collection Runner logic updated for iteration and data feed
*   [x] CSV data parsing implemented
*   [x] JSON data parsing implemented
*   [x] Data variables injected into request context for each iteration
*   [x] Manual tests performed
*   [x] Automated tests (where applicable) added/updated

### CODE REVIEW

File: packages/hoppscotch-common/src/helpers/import-export/dataset/parser.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        const currentLine = lines[i].split(",")
+
+        headers.forEach((header, j) => {
+            obj[header] = currentLine[j]?.trim() || ""
+        })
+
+        result.push(obj)
+    }
+
+    return result
```
[the remark this answers is not in the document]
Comment (Manohar-ford) [reply packages/hoppscotch-common/src/helpers/import-export/dataset/parser.ts]: This needs the papaparser to imported into the pnpm package so that it's based on the Hoppscoth need.

**Comment by Trishadring** [comment]
any update on when when can get implemented? this would same my team a ton of time :)

**Comment by broccolism** [comment]
Any progress here? My team is also eagerly waiting this feature. We seriously want to change thunder client to hoppscotch...

**Comment by rgrouhan** [comment]
Hello,
Can we hope to have this feature implemented soon ? Since postman changed this feature to a payed one, we're trying to find an alternative.

**Comment by chhaviG22** [comment]
hi @liyasthomas @jamesgeorge007 we're also looking for iteration and data file support in the Collection Runner, so this feature would be very useful for us. Is there any plan or timeline for getting this PR reviewed and released? Happy to help test it if that would be useful.

**Comment by sshadel-ftusa** [comment]
+1 would really like this feature in the client to fully replace Postman for our team

**Comment by nivedin** [comment]
Hi @Manohar-ford, thank you for this, and for your patience; this PR is what put iterations-data feed on our radar. We're moving forward with [#6513](https://github.com/hoppscotch/hoppscotch/pull/6513), which covers the same ground and is in review. The deciding factor is structural, this PR collects every iteration into one flat result list, so there's no per-iteration summary and no way to tell which iteration a failure came from, whereas #6513 treats each iteration as its own result set and scopes data values to the iteration, unwinding them afterwards so a run doesn't leave residue in the user's environment. Closing this one for that reason, not because the feature isn't wanted, review feedback from you on #6513 would be very welcome.
